examples/java8/src/main/java/org/apache/beam/examples/complete/game/utils/WriteToText.java

```diff
@@ -0,0 +1,193 @@
+import org.joda.time.format.DateTimeFormatter;
+
+/**
+ * Generate, format, and write rows. Use provided information about the field names and types, as
+ * well as lambda functions that describe how to generate their values.
+ */
+public class WriteToText<InputT>
+    extends PTransform<PCollection<InputT>, PDone> {
+
+  private static final DateTimeFormatter formatter =
+      DateTimeFormat.forPattern("yyyy-MM-dd HH:mm:ss.SSS")
+          .withZone(DateTimeZone.forTimeZone(TimeZone.getTimeZone("PST")));
+
+  protected String filenamePrefix;
+  protected Map<String, FieldInfo<InputT>> fieldInfo;
+  protected boolean windowed;
+
+  public WriteToText() {
+  }
+
+  public WriteToText(
+      String filenamePrefix,
+      Map<String, FieldInfo<InputT>> fieldInfo,
+      boolean windowed) {
+    this.filenamePrefix = filenamePrefix;
+    this.fieldInfo = fieldInfo;
+    this.windowed = windowed;
+  }
+
+  /**
+   * A {@link Serializable} function from a {@link DoFn.ProcessContext}
+   * and {@link BoundedWindow} to the value for that field.
+   */
+  public interface FieldFn<InputT> extends Serializable {
+    Object apply(DoFn<InputT, String>.ProcessContext context, BoundedWindow window);
+  }
+
+  /** Define a class to hold information about output table field definitions. */
+  public static class FieldInfo<InputT> implements Serializable {
+    // A lambda function to generate the field value
+    private FieldFn<InputT> fieldFn;
+
+    public FieldInfo(FieldFn<InputT> fieldFn) {
+      this.fieldFn = fieldFn;
+    }
+
+    FieldFn<InputT> getFieldFn() {
+      return this.fieldFn;
+    }
+  }
+  /** Convert each key/score pair into a row as specified by fieldFn. */
+  protected class BuildRowFn extends DoFn<InputT, String> {
+
+    @ProcessElement
+    public void processElement(ProcessContext c, BoundedWindow window) {
+      StringBuilder row = new StringBuilder();
+      for (Map.Entry<String, FieldInfo<InputT>> entry : fieldInfo.entrySet()) {
+        String key = entry.getKey();
+        FieldInfo<InputT> fcnInfo = entry.getValue();
+        FieldFn<InputT> fcn = fcnInfo.getFieldFn();
+        row.append(key);
+        row.append(": ");
+        row.append(fcn.apply(c, window));
+        row.append(", ");
+      }
+      String result = row.length() > 0 ? row.substring(0, row.length() - 2) : "";
+      c.output(result);
+    }
+  }
+
+  /**
+   * A {@link DoFn} that writes elements to files with names deterministically derived from the
+   * lower and upper bounds of their key (an {@link IntervalWindow}).
+   */
+  protected class WriteOneFilePerWindow extends PTransform<PCollection<String>, PDone> {
+
+    private final String filenamePrefix;
+
+    public WriteOneFilePerWindow(String filenamePrefix) {
+      this.filenamePrefix = filenamePrefix;
+    }
+
+    @Override
+    public PDone expand(PCollection<String> input) {
+      // filenamePrefix may contain a directory and a filename component. Pull out only the filename
```

Comment:
It would be polite to validate that `input.getWindowingStrategy().getWindowFn().windowCoder()` returns `IntervalWindow.getCoder()`. This is the Java SDK way of validating that the window type is compatible with the transform.